File: ios/Classes/AdropAdManager.swift
import Foundation
import Flutter
import AdropAds

class AdropAdManager: NSObject {
    private let messenger: FlutterBinaryMessenger
    private var ads: [String: AdropAd?] = [:]

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
    }

    func load(adType: AdType, unitId: String, requestId: String) {
        let ad = getAd(adType: adType, requestId: requestId)
        ?? createAd(adType: adType, unitId: unitId, requestId: requestId)
        let key = keyOf(adType, requestId)
        if self.ads[key] == nil {
            self.ads[key] = ad
        }

        ad?.load()
    }

    func show(adType: AdType, requestId: String) {
        if let ad = self.ads[keyOf(adType, requestId)] {
            ad?.show()
        }
    }

    func customize(adType: AdType, requestId: String, data: [String:Any]) {
        switch adType {
        case AdType.popup:
            guard let ad = self.ads[self.keyOf(adType, requestId)] as? FlutterAdropPopupAd else { return }

            ad.customize(data)
        default:
            return
        }
    }

    func destroy(adType: AdType, requestId: String) {
        let key = keyOf(adType, requestId)

        switch adType {
        case AdType.popup:
            guard let ad = self.ads[self.keyOf(adType, requestId)] as? FlutterAdropPopupAd else { return }
            
            DispatchQueue.main.async { [weak self] in
                ad.destroy()
                self?.removeAds(key)
            }
            
        default:
            removeAds(key)
        }
    }
    
    private func removeAds(_ key: String) {
        DispatchQueue.main.async { [weak self] in
            self?.ads.removeValue(forKey: key)
        }
    }

    func createAd(adType: AdType, unitId: String, requestId: String) -> AdropAd? {
        switch adType {
        case .interstitial:
            return FlutterAdropInterstitialAd(unitId: unitId, requestId: requestId, messenger: messenger)
        case .rewarded:
            return FlutterAdropRewardedAd(unitId: unitId, requestId: requestId, messenger: messenger)
        case .popup:
            return FlutterAdropPopupAd(unitId: unitId, requestId: requestId, messenger: messenger)
        case .native:
            return FlutterAdropNativeAd(unitId: unitId, requestId: requestId, messenger: messenger)
        case .undefined:
            return nil
        }
    }

    func getAd(adType: AdType, requestId: String) -> AdropAd? {
        guard let ad = self.ads[keyOf(adType, requestId)] else {
            return nil
        }

        return ad

    }

    private func keyOf(_ adType: AdType, _ requestId: String) -> String {
        return "\(adType)/\(requestId)"
    }
}

File: android/src/main/kotlin/io/adrop/adrop_ads/AdropAdManager.kt
package io.adrop.adrop_ads

import android.app.Activity
import android.content.Context
import io.adrop.adrop_ads.popupAd.FlutterAdropPopupAd
import io.adrop.adrop_ads.interstitial.FlutterAdropInterstitialAd
import io.adrop.adrop_ads.rewarded.FlutterAdropRewardedAd
import io.flutter.plugin.common.BinaryMessenger

class AdropAdManager {

    private val ads: MutableMap<String, AdropAd?> = mutableMapOf()

    fun load(context: Context, adType: AdType, unitId: String, requestId: String, messenger: BinaryMessenger) {
        val ad = getAd(adType, requestId) ?: createAd(context, adType, unitId, requestId, messenger)
        val key = keyOf(adType, requestId)
        ads[key]?:let {
            ads[key] = ad
        }
        ad?.load()
    }

    fun show(adType: AdType, requestId: String, activity: Activity) {
        getAd(adType, requestId)?.show(activity)
    }

    fun customize(adType: AdType, requestId: String, data: Map<String, Any>) {
        when (adType) {
            AdType.Popup -> {
                val popupAd = ads[keyOf(adType, requestId)] as? FlutterAdropPopupAd
                popupAd?: run { return }

                popupAd.customize(data)
            }
            else -> return
        }
    }

    fun destroy(adType: AdType, requestId: String) {
        val key = keyOf(adType, requestId)
        ads[key]?.let {
            it.destroy()
            ads.remove(key)
        }
    }

    private fun createAd(context: Context, adType: AdType, unitId: String, requestId: String, messenger: BinaryMessenger): AdropAd? {
        return when (adType) {
            AdType.Interstitial -> FlutterAdropInterstitialAd(context, unitId, requestId, messenger)
            AdType.Rewarded -> FlutterAdropRewardedAd(context, unitId, requestId, messenger)
            AdType.Popup -> FlutterAdropPopupAd(context, unitId, requestId, messenger)
            AdType.Undefined -> null
        }
    }

    private fun getAd(adType: AdType, requestId: String): AdropAd? {
        return ads[keyOf(adType, requestId)]
    }

    private fun keyOf(adType: AdType, requestId: String): String = "$adType/$requestId"

}
